godtossing/Pos/Ssc/GodTossing/Toss/Base.hs

```diff
@@ -144,78 +227,50 @@ computeSharesDistrPure
+    coins = map (fromIntegral . unsafeGetCoin . snd) (HM.toList richmen)
+
+    portions :: [Double]
+    portions = map ((/ fromIntegral totalCoins) . fromIntegral) coins
+
+    totalCoins :: CoinUnsafe
+    totalCoins = sum coins
+
     -- We multiply all portions by mult and divide them on their gcd
     --     we get commitment distribution
     -- compute sum difference between real portions and current portions
     -- select optimum using next strategy:
     --   * if sum error < epsilon - we try minimize sum of commitments
     --   * otherwise we try minimize sum error
-    compute fromX toX epsilon portions = do
-        for_ [fromX..toX] $ \x -> do
-            let curDistrN = multPortions portions x
-            when (all (> 0) curDistrN) $ do
-                let curDistr = normalize curDistrN
-                let s = sum curDistr
-                let curPortions = map (`divRat` s) curDistr
-                let delta = calcSumError curPortions portions
-                (optDelta, optS, _) <- get
-                -- if delta less than epsilon then we try minimize sum of commitments
-                when (delta <= epsilon && s < optS ||
-                     -- otherwise we try minimize sum error of rounding
-                      delta <= optDelta && delta > epsilon) $
-                      put (delta, s, curDistr)
-
-    calcSumError:: [Rational] -> [Rational] -> Rational
-    calcSumError pNew p = do
-        let sorted1 = sortOn (\(a, b) -> b / a) $ zip p pNew
-        let sorted2 = sortOn (\(a, b) -> b - a) $ zip p pNew
-        let res1 = max (computeError1 0 0 sorted1) (computeError2 0 0 $ reverse sorted1)
-        let res2 = max (computeError1 0 0 sorted2) (computeError2 0 0 $ reverse sorted2)
-        max res1 res2
-
-    half = 0.5::Rational
-    -- Error when real stake more than 0.5 but our new stake less
-    computeError1 _ _ [] = 0
-    computeError1 real new (x:xs)
-        | new < half && real >= half =
-            max (real - half) $ computeError1 (real + fst x) (new + snd x) xs
-        | otherwise = computeError1 (real + fst x) (new + snd x) xs
-
-    -- Error when real stake less than 0.5 but our new stake more
-    computeError2 _ _ [] = 0
-    computeError2 real new (x:xs)
-        | new >= half && real < half =
-            max (half - real) $ computeError2 (real + fst x) (new + snd x) xs
-        | otherwise = computeError2 (real + fst x) (new + snd x) xs
-
-    multPortions :: [Rational] -> Word16 -> [Word16]
-    multPortions p (toRational -> mult) = map (truncate . (* mult)) p
-
-    divRat :: (Real a, Real b) => a -> b -> Rational
-    divRat x y = toRational x / toRational y
+    compute :: Word16 -> Word16 -> Word16 -> Maybe [Word16]
+    compute !x !toX !prevSum
+        | x > toX = Nothing
+        | otherwise = do
+            let curDistrN = multPortions x
+            let curDistr = normalize curDistrN
+            let !s = sum curDistr
+            if s == prevSum then compute (x + 1) toX prevSum
```

Comment:
1. it is. when you add `1` to x you add to each shares distribution `i` real `p_i`, and if each number isn't exceeded next integer number then each number equals itself previous value - a sum of distributions is same 
2. no, it's not possible. because >each number equals itself previous value